cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -531,65 +532,117 @@ private MetadataEntry convertRow(Row row) {
       }
       try {
         Row next;
         while ((next = scanner.next()) != null) {
-          String rowValue = next.getString(indexColumn);
-          if (rowValue == null) {
-            continue;
-          }
-
-          final byte[] rowKey = next.getRow();
-          String targetType = MdsKey.getTargetType(rowKey);
+          processRow(results, next, DEFAULT_INDEX_COLUMN, types);
+        }
+      } finally {
+        scanner.close();
+      }
+    }
+    return new SearchResults(results, Collections.<String>emptyList());
+  }
 
-          // Filter on target type if not set to include all types
-          if (!includeAllTypes &&
-            !types.contains(MetadataSearchTargetType.valueOfSerializedForm(targetType))) {
+  private SearchResults searchByCustomIndex(String namespaceId, Set<MetadataSearchTargetType> types,
+                                            SortInfo sortInfo, int offset, int limit, int numCursors,
+                                            @Nullable String cursor) {
+    List<MetadataEntry> results = new ArrayList<>();
+    String indexColumn = getIndexColumn(sortInfo.getSortBy(), sortInfo.getSortOrder());
+    int fetchSize = offset + numCursors * limit;
+    List<String> cursors = new ArrayList<>(numCursors);
+    for (String searchTerm : getSearchTerms(namespaceId, "*")) {
+      byte[] startKey = Bytes.toBytes(searchTerm.substring(0, searchTerm.lastIndexOf("*")));
+      byte[] stopKey = Bytes.stopKeyForPrefix(startKey);
+      // if a cursor is provided, then start at the cursor
+      if (!Strings.isNullOrEmpty(cursor)) {
+        String namespaceInStartKey = searchTerm.substring(0, searchTerm.indexOf(KEYVALUE_SEPARATOR));
+        startKey = Bytes.toBytes(namespaceInStartKey + KEYVALUE_SEPARATOR + cursor);
+      }
+      try (Scanner scanner = indexedTable.scanByIndex(Bytes.toBytes(indexColumn), startKey, stopKey)) {
+        Row next;
+        int count = 0;
+        while ((next = scanner.next()) != null && count < fetchSize) {
+          // skip until we reach offset
+          if (count++ < offset) {
             continue;
           }
-
-          NamespacedEntityId targetId = MdsKey.getNamespacedIdFromKey(targetType, rowKey);
-          String key = MdsKey.getMetadataKey(targetType, rowKey);
-          MetadataEntry entry = getMetadata(targetId, key);
-          results.add(entry);
+          processRow(results, next, indexColumn, types);
+          if (results.size() > limit) {
+            // A cursor is the first element of the a chunk of ordered results. Since its always the first element,
+            // we want to add a key as a cursor, if upon dividing the current number of results by the chunk size,
+            // the remainder is 1. However, this is not true, when the chunk size is 1, since in that case, the
+            // remainder on division can never be 1, it is always 0.
+            int mod = limit == 1 ? 0 : 1;
```

Comment:
nit: better to check mod first, since it will result in fewer overall comparisons.